src/main/webapp/app/shared/modal/AddMutationModal.tsx

```diff
@@ -80,507 +82,747 @@ function AddMutationModalContent({
+        fetchExcludedAlteration(newValue, tabIndex, excludingIndex, tabStates);
+      }
+    } else {
+      if (isDebounced) {
+        handleNormalAlterationChange(newValue, tabIndex);
+      } else {
+        fetchNormalAlteration(newValue, tabIndex, tabStates);
+      }
+    }
+  }
+
+  function handleNormalFieldChange(newValue: string, field: keyof AlterationData, tabIndex: number) {
+    setTabStates(states => {
+      const newStates = _.cloneDeep(states);
+      newStates[tabIndex][field as string] = newValue;
+      return newStates;
+    });
+  }
+
+  function handleExcludingFieldChange(newValue: string, field: keyof AlterationData, tabIndex: number, excludingIndex: number) {
+    setTabStates(states => {
+      const newStates = _.cloneDeep(states);
+      newStates[tabIndex].excluding[excludingIndex][field as string] = newValue;
+      return newStates;
+    });
+  }
+
+  function handleFieldChange(newValue: string, field: keyof AlterationData, tabIndex: number, excludingIndex?: number) {
+    notNullOrUndefined(excludingIndex)
+      ? handleExcludingFieldChange(newValue, field, tabIndex, excludingIndex)
+      : handleNormalFieldChange(newValue, field, tabIndex);
+  }
+
   const handleKeyDown: KeyboardEventHandler = async event => {
     if (!inputValue) return;
     if (event.key === 'Enter' || event.key === 'tab') {
-      await fetchAlteration(inputValue);
+      const newParsedAlteration = filterAlterationsAndNotify(parseAlterationName(inputValue), tabStates);
+      if (newParsedAlteration.length === 0) {
+        return;
+      }
+
+      const newComment = newParsedAlteration[0].comment;
+      const newVariantName = newParsedAlteration[0].name;
+
+      const newEntityStatusAlterationsPromise = fetchAlterations(newParsedAlteration.map(alt => alt.alteration));
+      const newEntityStatusExcludingAlterationsPromise = fetchAlterations(newParsedAlteration[0].excluding);
+      const [newEntityStatusAlterations, newEntityStatusExcludingAlterations] = await Promise.all([
+        newEntityStatusAlterationsPromise,
+        newEntityStatusExcludingAlterationsPromise,
+      ]);
+
+      const newExcludingAlterations = newEntityStatusExcludingAlterations.map((alt, index) =>
+        convertEntityStatusAlterationToAlterationData(alt, newParsedAlteration[0].excluding[index], [], '')
+      );
+      const newAlterations = newEntityStatusAlterations.map((alt, index) =>
+        convertEntityStatusAlterationToAlterationData(
+          alt,
+          newParsedAlteration[index].alteration,
+          _.cloneDeep(newExcludingAlterations),
+          newComment,
+          newVariantName
+        )
+      );
+
+      setTabStates(states => [...states, ...newAlterations]);
 
       setInputValue('');
       event.preventDefault();
     }
   };
 
-  function getTabTitle(mutationState: MutationState) {
-    if (!mutationState) {
+  const handleKeyDownExcluding = async (event: React.KeyboardEvent<Element>, tabIndex: number) => {
```

Comment:
Isn't this method almost identical with `handleKeyDown`?

Reply:
I think there are enough differences to justify two different methods. The `setState` calls are different, `handleKeyDown` needs to fetch excluding and non-excluding while `handleKeyDownExcluding` only needs to fetch its own alterations, and `handleKeyDownExcluding` handles checking for duplicate alterations differently.

I think there are a lot of cases where the logic for normal alterations and excluding alterations appear similar, but they are just different enough to make abstracting the logic messy. I tried to abstract a bunch the first time I created this component, and I think the resulting code was harder to maintain.